### PR TITLE
Fee-market in EVM

### PR DESCRIPTION
**Pull request by hujw77**

### CODE REVIEW

**Comment by hujw77** [comment]
Thanks to @AsceticBear for good explanation at `fee-market`
```
account R1-> Fee market -> call `enroll_and_lock_collateral(collateral, relay_fee)` -> item(<R1(collater, relay_fee), R2(..), R3(..)>)
1. collater: check with usable balance,   
2. relay_fee: check with a fixed value.

CollateralPerOrder = 10, R1 collater: 50 -> R1 5 order capacity -> m1, m2, m3, m4, m5, m6

-> m1, m2, m3 confirm 

- `update_collateral`: update order capacity -> decrease collateral -> CollateralPerOrder = 10, R1 collater: 50 ->  m1, m2, m3 -> order capacity = 2

Select assigned relayer list:
- all relayers who has extra order capacity, relay fee sort, 3 assigned relayer -> third relayer's fee -> market fee

New order process:
- user -> s2s message(A -> B) -> fee market create a order(sent_time, 3 个 assigned relayers) -> store

r1 -> r2 -> r3
p1 -> p2 -> p3(market_fee)

pub struct Order<AccountId, BlockNumber, Balance> {
	pub lane: LaneId,
	pub message: MessageNonce,
	pub sent_time: BlockNumber,
	pub confirm_time: Option<BlockNumber>,
	pub locked_collateral: Balance,
	pub relayers: Vec<PriorRelayer<AccountId, BlockNumber, Balance>>,
}

pub struct PriorRelayer<AccountId, BlockNumber, Balance> {
	pub id: AccountId,
	pub fee: Balance,
	pub valid_range: Range<BlockNumber>,
}

order sent_time: 100, slot = 50 => Order relayer 1: [100-150] relayer2: [150-200] relayer3: [200-250]

Confirm order process:
-> call back -> query order storage -> set order's confirm_time -> store


Reward part(payment.rs): confirm_time: [180] -> r2

60% p1 -> assigned relayer r2
80% * (40% * p1) -> message relayer
20% * (40% * p1) -> confirm relayer
p3 - p1 -> treasury

Slash part(payment.rs): confirm_time: [300]

slash_value = max(2 coin * 50 = 100 coin, locked_collateral) + p3.fee

80% slash_value -> message relayer
20% slash_value -> confirm relayer
```

File: contracts/bridge/contracts/common/fee-market/FeeMarket.sol [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+                index++;
+            } else {
+                prune_relayer(prev, cur);
+            }
+            prev = cur;
+            cur = relayers[cur];
+        }
+        require(index == ASSIGNED_RELAYERS_NUMBER, "!assigned");
+        return array;
+    }
```
Comment (HackFisher) [inline contracts/bridge/contracts/common/fee-market/FeeMarket.sol:268]: simplify implementation and remove `prune_relayer`.

`balanceOf[cur]` should be sorted.

File: contracts/bridge/contracts/common/fee-market/FeeMarket.sol [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+
+    function _get_and_prune_top_relayers() internal returns (address[] memory) {
+        require(ASSIGNED_RELAYERS_NUMBER <= relayer_count, "!count");
+        address[] memory array = new address[](ASSIGNED_RELAYERS_NUMBER);
+        uint index = 0;
+        address prev = SENTINEL_HEAD;
+        address cur = relayers[SENTINEL_HEAD];
+        while (cur != SENTINEL_TAIL && index < ASSIGNED_RELAYERS_NUMBER) {
+            if (balanceOf[cur] >= COLLATERAL_PERORDER) {
```
Comment (HackFisher) [inline contracts/bridge/contracts/common/fee-market/FeeMarket.sol:257]:
```suggestion
            if (balanceOf[cur] - lockedOf[cur] >= COLLATERAL_PERORDER) {
```

Should be available part.

File: contracts/bridge/contracts/common/fee-market/FeeMarket.sol (hunk 1 above)
Comment (hujw77) [reply contracts/bridge/contracts/common/fee-market/FeeMarket.sol:268]: `feeOf[src]` is sorted, not `balanceOf[cur]`

File: contracts/bridge/contracts/common/fee-market/FeeMarket.sol [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+    function _remove_relayer(address prev, address cur) private {
+        require(cur != address(0) && cur != SENTINEL_HEAD && cur != SENTINEL_TAIL, "!valid");
+        require(relayers[prev] == cur, "!cur");
+        require(lockedOf[cur] == 0, "!locked");
+        relayers[prev] = relayers[cur];
+        relayers[cur] = address(0);
+        feeOf[cur] = 0;
+        relayer_count--;
+        emit RemoveRelayer(prev, cur);
+    }
```
Comment (HackFisher) [inline contracts/bridge/contracts/common/fee-market/FeeMarket.sol:279]: let user remove them self.

File: contracts/bridge/contracts/common/fee-market/FeeMarket.sol [hunk 4]
```diff
@@ -0,0 +1,416 @@
...
+        withdraw(balanceOf[msg.sender]);
+        remove_relayer(prev);
+    }
+
+    // enroll to be a relayer
+    // `prev` is the previous relayer
+    // `fee` is the maker fee to set, PrevFee <= MakerFee <= NextFee
+    function add_relayer(address prev, uint fee) public enoughBalance {
+        address cur = msg.sender;
+        address next = relayers[prev];
```
Comment (HackFisher) [inline contracts/bridge/contracts/common/fee-market/FeeMarket.sol:193]:
```suggestion
        address next = relayers[prev + 1];
```
I'm not sure whether it should be `prev + 1` here.

File: contracts/bridge/contracts/common/fee-market/FeeMarket.sol [hunk 5]
```diff
@@ -0,0 +1,416 @@
...
+                uint256 base_fee = feeOf[assign_relayer];
+                (delivery_reward, confirm_reward, vault_reward) = _distribute_ontime(message_fee, base_fee, assign_relayer, delivery_relayer, confirm_relayer);
+            }
+            _unlock(assign_relayer, COLLATERAL_PERORDER);
+        }
+    }
+
+    function _slash_and_unlock_late(uint256 key, uint256 diff_time) internal returns (uint256 delivery_reward, uint256 confirm_reward) {
+        uint256 message_fee = getOrderFee(key);
+        uint256 late_time = diff_time - ASSIGNED_RELAYERS_NUMBER * RELAY_TIME;
```
Comment (HackFisher) [inline contracts/bridge/contracts/common/fee-market/FeeMarket.sol:369]: pass in late_time instead.

File: contracts/bridge/contracts/common/fee-market/FeeMarket.sol [hunk 6]
```diff
@@ -0,0 +1,416 @@
...
+    // System treasury
+    address public immutable VAULT;
+    // The collateral relayer need to lock for each order.
+    uint256 public immutable COLLATERAL_PERORDER;
+    // Fee market assigned relayers numbers
+    uint256 public immutable ASSIGNED_RELAYERS_NUMBER;
+    // SLASH_AMOUNT = COLLATERAL_PERORDER * late_time / SLASH_TIME
+    uint256 public immutable SLASH_TIME;
+    // Time assigned relayer to relay messages
+    uint256 public immutable RELAY_TIME;
```
Comment (HackFisher) [inline contracts/bridge/contracts/common/fee-market/FeeMarket.sol:32]: Configurable by owner, and record locked_collateral and assigned_relayer_count in order.

**Review by HackFisher — COMMENTED** [review]
LGTM so far, and requested some changes.

File: contracts/bridge/contracts/common/fee-market/FeeMarket.sol [hunk 7]
```diff
@@ -0,0 +1,448 @@
...
+    address public immutable VAULT;
+
+    // SlashAmount = CollateralPerorder * LateTime / SlashTime
+    uint32 public SlashTime;
+    // Time assigned relayer to relay messages
+    uint32 public RelayTime;
+    // Fee market assigned relayers numbers
+    uint32 public AssignedRelayersNumber;
+    // The collateral relayer need to lock for each order.
+    uint256 public CollateralPerorder;
```
Comment (HackFisher) [inline contracts/bridge/contracts/common/fee-market/FeeMarket.sol:34]: naming, start with lowercase.

File: contracts/bridge/contracts/common/fee-market/FeeMarket.sol [hunk 8]
```diff
@@ -0,0 +1,448 @@
...
+        uint32 assigned_time;
+        // Assigned number of relayers
+        uint32 assigned_relayers_number;
+        // Assigned collateral of each relayer
+        uint256 collateral;
+    }
+    // message_encoded_key => Order
+    mapping(uint256 => Order) public orderOf;
+    // message_encoded_key => assigned_slot => assigned_relayer
+    mapping(uint256 => mapping(uint256 => address)) public assigned_relayers;
```
Comment (HackFisher) [inline contracts/bridge/contracts/common/fee-market/FeeMarket.sol:62]: https://docs.soliditylang.org/en/v0.5.3/style-guide.html#naming-conventions

Not quite sure about naming conventions, but it's a good time to do clean up now.

File: contracts/bridge/contracts/common/fee-market/FeeMarket.sol (hunk 7 above)
Comment (hujw77) [reply contracts/bridge/contracts/common/fee-market/FeeMarket.sol:34]: `camelCase` is good for storage variables

File: contracts/bridge/contracts/common/fee-market/FeeMarket.sol (hunk 8 above)
Comment (hujw77) [reply contracts/bridge/contracts/common/fee-market/FeeMarket.sol]: There I use `snake_case` for `order` is a good distinction for general storage variables

Comment (hujw77) [reply contracts/bridge/contracts/common/fee-market/FeeMarket.sol]: For function.
`snake_case` function is  for relayer to call.
`camelCase` function is for all user to call.

**Review by HackFisher — APPROVED** [review]

**Pull request merged**